demos/forward.py

```diff
@@ -45,31 +46,75 @@
 
 g_verbose = True
 
+class ForwardServer(SocketServer.ThreadingTCPServer,object):
+    """
+    Use ForwardServer in two ways:
 
-class ForwardServer (SocketServer.ThreadingTCPServer):
-    daemon_threads = True
-    allow_reuse_address = True
-    
+    1. Call run() directly to start the forwarding.
+
+        ForwardServer(...).run()
+
+       In this usage pattern, the thread calling run() will be blocked until stop() is called from another thread.
+
+    2. Use "with" statement to start server in a daemon thread.
+
+        with ForwardServer(...) as serv:
+            # Use the forwarded port
+
+       In this usage pattern, the thread creating the server is not blocked and so can do something useful with the forwarded port.
+       The server will be automatically stopped at the end of the with block.
+    """
+    def __init__(self, local_port, remote_host, remote_port, ssh_transport):
+        # Save these for use by Handler class.
+        self.remote_host = remote_host
+        self.remote_port = remote_port
+        self.ssh_transport = ssh_transport
+        super(ForwardServer, self).__init__(('', local_port), Handler)
+        self.daemon_threads = True
+        self.allow_reuse_address = True
+        self.th = None
+
+    def run(self):
+        self.serve_forever()
 
-class Handler (SocketServer.BaseRequestHandler):
+    def stop(self):
+        self.shutdown()
+        self.socket.close()
 
+    def join(self):
+        if self.th:
+            # Workaround for Thread.join() prevents KeyboardInterrupt from getting raised.
+            while self.th.isAlive():
```

Comment:
Please use the newer ```is_alive```, there is no difference, but we are trying to keep the code base modern and forward compatible. :)